external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCSpout.java

```diff
@@ -0,0 +1,530 @@
+            // short-cut: return buffer-internal array
+            return buf.array();
+        }
+        // copy buffers into result byte[]
+        byte[] arr = new byte[read];
+        int pos = 0;
+        for (ByteBuffer b : bufs) {
+            r = b.remaining();
+            b.get(arr, pos, r);
+            pos += r;
+        }
+        buf.get(arr, pos, buf.remaining());
+        return arr;
+    }
+
+    private static String httpHeadersVerbatim(HttpMessage http) {
+        return new String(http.serializeHeader(), StandardCharsets.UTF_8);
+    }
+
+    private void addVerbatimHttpHeaders(Metadata metadata,
+            WarcResponse response, HttpResponse http, HttpRequest request) {
+        metadata.addValue(protocolMDprefix + ProtocolResponse.REQUEST_TIME_KEY,
+                Long.toString(response.date().toEpochMilli()));
+        if (response.ipAddress().isPresent()) {
+            metadata.addValue(
+                    protocolMDprefix + ProtocolResponse.RESPONSE_IP_KEY,
+                    response.ipAddress().get().getHostAddress());
+        }
+        if (request != null) {
+            metadata.addValue(
+                    protocolMDprefix + ProtocolResponse.REQUEST_HEADERS_KEY,
+                    httpHeadersVerbatim(request));
+        }
+        metadata.addValue(
+                protocolMDprefix + ProtocolResponse.RESPONSE_HEADERS_KEY,
+                httpHeadersVerbatim(http));
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    @Override
+    public void open(Map conf, TopologyContext context,
+            SpoutOutputCollector collector) {
+        _collector = collector;
+        record = Optional.empty();
+
+        maxContentSize = ConfUtils.getInt(conf, "http.content.limit", -1);
+        if (contentBufferSize > maxContentSize) {
+            // no need to buffer more content than max. used
+            contentBufferSize = maxContentSize;
+        }
+        storeHTTPHeaders = ConfUtils.getBoolean(conf, "http.store.headers",
+                false);
+        protocolMDprefix = ConfUtils.getString(conf,
+                ProtocolResponse.PROTOCOL_MD_PREFIX_PARAM, protocolMDprefix);
+        sleepEmitFetched = ConfUtils.getLong(conf,
+                "warc.spout.emit.fetched.sleep.ms", 0);
+
+        int metricsTimeBucketSecs = ConfUtils.getInt(conf,
+                "fetcher.metrics.time.bucket.secs", 10);
+        eventCounter = context.registerMetric("warc_spout_counter",
+                new MultiCountMetric(), metricsTimeBucketSecs);
+    }
+
+    @Override
+    public void nextTuple() {
+        if (!active)
+            return;
+
+        if (buffer.isEmpty()) {
+            try {
+                populateBuffer();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        if (warcReader == null && buffer.isEmpty()) {
+            // input exhausted
+            try {
+                Thread.sleep(10);
```

Comment:
don't need it -  Storm will handle sleeping between calls to nextTuple if you don't emit anything during a call

Reply:
Ok, will remove it - was following the API doc of [ISpout::nextTuple](https://storm.apache.org/releases/1.2.3/javadocs/org/apache/storm/spout/ISpout.html#nextTuple--), obviously outdated.

Reply:
Let me double check in Storm's code

Reply:
see http://storm.apache.org/releases/current/Performance.html -> wait strategy

default values 

_topology.sleep.spout.wait.strategy.time.ms | 1
topology.spout.wait.strategy | "org.apache.storm.spout.SleepSpoutWaitStrategy"_

Reply:
Ok: sleep removed.